src/pdn/src/renderer.cpp

```diff
@@ -125,6 +119,11 @@ void PDNRenderer::update()
     }
   }
 
+  shapes_ = Shape::convertVectorToTree(shapes);
+  grid_obstructions_ = Shape::convertVectorToTree(obs);
+  shapes_ = Shape::convertVectorToTree(shapes);
+  vias_ = Shape::convertVectorToTree(vias);
```

Comment:
Need to clear/shrink as you go to minimize the memory peak (here or in convertVectorToTree)